Add the `v0.3.0` release post (#200)
posit-dev/great-tables: docs/blog/introduction-0.3.0/index.qmd

---

> title: "Great Tables `v0.3.0`: So Many More Options!"

Reviewer: I have wrapped up adding review comments and just want to say that this is a really great looking post!

html-table-processing: none
author: Rich Iannone
date: 2024-02-16
freeze: true
jupyter: python3
---

As our work on **Great Tables** continues, we want you to be able to produce increasingly sophisticated tables. The look of an HTML table really matters and we believe aesthetics can elevate the presentation of tabular data. In the `v0.3.0` release, we've implemented features that are concerned with modifying the visual aspects of a table. Let's get down to what's new in this version.

### Modifying the widths of columns

Before `v0.3.0`, you could not alter the widths of individual columns. This meant that to great extent your content decided the width of individual columns. Even though browsers do an adequate job in sizing the widths of table columns, it doesn't always result in a pleasing-to-look-at table. What if you want more space? Maybe you want consistently-sized columns? There's many reasons to want to have a choice in the matter and the new `cols_width()` method now makes this possible. Here's an example of a table where the column widths are defined with `cols_width()`, using a mix of pixel and percentage widths applied to select columns.

> ```{python}

Reviewer: I'm not sure what I'm supposed to be looking for in this example's output. I wonder if it might be easier for readers to spot the effect with only the example that sets px widths for everything?

The API reference page has this example, so a nice move might be pointing people to that at the end of the section? (Added as a suggestion below)

Author: This is greatly simplified now and points to the reference page. Thanks!

from great_tables import GT, exibble

exibble_mini = exibble[["num", "char", "date", "datetime", "row"]].head(5)

(
    GT(exibble_mini)
    .cols_width(
        cases={
            "num": "150px",
            "char": "10%",
            "date": "30%"
        }
    )
)
```

In the previous example, the width settings govern the preferred sizes of the columns. But, depending on the nature of the content, the browser has license to make adjustments. An interesting thing happens if we set the width of all columns with pixel values: the table must use the specified widths as provided. In this next example we'll set widths for all columns, ensuring that the widths you specify are fully respected.

```{python}
(
    GT(exibble_mini)
    .cols_width(
        cases={
            "num": "30px",
            "char": "100px",
            "date": "100px",
            "datetime": "200px",
            "row": "50px"
        }
    )
)
```

> 

Reviewer: ```suggestion
See the [cols_width()](`great_tables.GT.cols_width`) API reference page for information.

```

```{python}
gt_tbl.opt_align_table_header(align="left")
```

With that, you don't have to hunt through the myriad options within `tab_options()` to find the two args you need to get the job done.

The `opt_all_caps()` method transforms the text within the column labels, the stub, and in all row groups so that we get an all-capitalized (yet somewhat sized down) look that better differentiates the labels from the data. It's rather easy to use, just do this:

```{python}
gt_tbl.opt_all_caps()
```

This sets nine options you'd otherwise set in `tab_options()` all at once, making life generally easier.

Here's one last example, this time using `opt_vertical_padding()`. You'd use that if you're dissatisfied with the level of top/bottom padding within cells of all locations (e.g., in the table body, in the column labels, etc.). You can either make a table taller or more 'compressed' with a single argument: `scale=`. Here's an example where the amount of vertical padding is reduced, resulting in a table taking up less vertical space.

```{python}
gt_tbl.opt_vertical_padding(scale=0.5)
```

### A new formatting method: `fmt_image()`

> Wouldn't it be great to add graphics to your table? I mean, you can already do that with `fmt_markdown()` by creating a link to an image resource, but `fmt_image()` provides an easier way to add image files on disk into table body cells. This method allows for one or more images to be placed in the targeted cells. The cells need to contain some reference to an image file, either: (1) complete http/https or local paths to the files; (2) the file names, where a common path can be provided via the `path=` arg; or (3) a fragment of the file name, where the `file_pattern=` arg helps to compose the entire file name and `path=` provides the path information.

Reviewer: ```suggestion
Wouldn't it be great to add graphics to your table? `fmt_image()` provides an easy way to add image files on disk into table body cells. The cells need to contain some reference to an image file. The `path=` and `file_pattern=` arguments give you some flexibility in defining exactly where the image files live.
```

(Note that I made another suggestion below the example with the deleted text!)

Here's an example using the `metro` dataset that's included within **Great Tables**.

```{python}
from great_tables.data import metro
from importlib_resources import files

img_paths = files("great_tables") / "data/metro_images"
metro_mini = metro[["name", "lines", "passengers"]].head(5)

(
    GT(metro_mini)
    .fmt_image(columns="lines", path=img_paths, file_pattern="metro_{}.svg")
    .fmt_integer(columns="passengers")
    .cols_label(
        name="Station",
        lines="Metro Lines",
        passengers="Passengers per Year (2021)"
    )
    .tab_options(table_width="700px")
)
```

> 

Reviewer: ```suggestion

Notice that `path=img_paths` specified the folder the images live in, and `file_pattern="metro_{}.svg"` provided a template for converting each value in the lines column to an svg file name.

`fmt_image()` supports 3 kinds of files as inputs, either: (1) complete http/https or local paths to the files; (2) the file names, where a common path can be provided via the `path=` arg; or (3) a fragment of the file name, as shown in the example above.
```

The package has some graphics stored in the `data/metro_images` directory. They are SVGs are they look *very* nice in this table!

> 

Reviewer: ```suggestion
See the [fmt_image()](`great_tables.GT.fmt_image`) reference page for more information.

```
